configuration/src/main/java/io/airlift/configuration/secrets/SecretsPluginManager.java

```diff
@@ -0,0 +1,205 @@
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Streams.stream;
+import static io.airlift.configuration.ConfigurationUtils.replaceEnvironmentVariables;
+import static java.nio.file.Files.newDirectoryStream;
+import static java.util.Arrays.asList;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public final class SecretsPluginManager
+{
+    private static final ImmutableList<String> SPI_PACKAGES = ImmutableList.<String>builder()
+            .add("io.airlift.spi.secrets")
+            .build();
+
+    private static final Logger log = Logger.get(SecretsPluginManager.class);
+
+    private static final String SECRETS_PROVIDER_NAME_PROPERTY = "secrets-provider.name";
+
+    private static final Pattern CONFIG_PROVIDER_NAME_PATTERN = Pattern.compile("[a-zA-Z][a-zA-Z0-9_-]*");
+
+    private final Map<String, SecretProviderFactory> secretsProviderFactories = new ConcurrentHashMap<>();
+    private final AtomicReference<Map<String, SecretProvider>> secretsProviders = new AtomicReference<>(ImmutableMap.of());
+    private final File installedSecretPluginsDir;
+    private final TomlConfiguration tomlConfiguration;
+
+    public SecretsPluginManager(TomlConfiguration tomlConfiguration)
+    {
+        this.tomlConfiguration = requireNonNull(tomlConfiguration, "tomlConfiguration is null");
+        ConfigurationFactory configurationFactory = new ConfigurationFactory(tomlConfiguration.getParentConfiguration());
+        SecretsPluginConfig config = configurationFactory.build(SecretsPluginConfig.class);
+        this.installedSecretPluginsDir = config.getSecretsPluginsDir();
+    }
+
+    public void installPlugins()
+    {
+        installSecretsPlugin(new EnvironmentVariableSecretsPlugin());
+
+        listFiles(installedSecretPluginsDir).stream()
+                .filter(File::isDirectory)
+                .forEach(file -> loadConfigurationResolvers(file.getAbsolutePath(), () -> createClassLoader(file.getName(), buildClassPath(file))));
+    }
+
+    @VisibleForTesting
+    void installSecretsPlugin(SecretsPlugin secretsPlugin)
+    {
+        secretsPlugin.getSecretProviderFactories()
+                .forEach(this::addSecretProviderFactory);
+    }
+
+    private void addSecretProviderFactory(SecretProviderFactory secretProviderFactory)
+    {
+        verify(CONFIG_PROVIDER_NAME_PATTERN.matcher(secretProviderFactory.getName()).matches(), "Secret provider name '%s' doesn't match pattern '%s'", secretProviderFactory.getName(), CONFIG_PROVIDER_NAME_PATTERN);
```

Comment:
But it would break with the `${ENV:...}` which we apply today right ?

Reply:
Are you sure it would break that.  The matcher is only used to validate the name of the factory. All names are lowered, so I assume this is all case insensitive.

Reply:
Sorry I was confused with the patter which matches the `${ENV:...}` - You are right - we could modify it to work only with lower case